scripts/genpinctrl/genpinctrl.py

```diff
@@ -0,0 +1,474 @@
+"""
+Utility to autogenerate Zephyr DT pinctrl files for all STM32 microcontrollers.
+
+Usage::
+
+    python genpinctrl.py -p /path/to/STM32CubeMX -o /path/to/output_dir
+
+Copyright (c) 2020 Teslabs Engineering S.L.
+
+SPDX-License-Identifier: Apache-2.0
+"""
+
+import argparse
+from collections import OrderedDict
+import glob
+import logging
+import os
+import re
+import shutil
+import xml.etree.ElementTree as ET
+
+from jinja2 import Environment, FileSystemLoader
+import yaml
+
+
+logger = logging.getLogger(__name__)
+
+
+SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
+"""Script directory."""
+
+TEMPLATE_FILE = "pinctrl-template.j2"
+"""pinctrl template file."""
+
+DEF_CONF_FILE = os.path.join(SCRIPT_DIR, "config.yaml")
+"""Default configuration file."""
+
+NS = "{http://mcd.rou.st.com/modules.php?name=mcu}"
+"""MCU XML namespace."""
+
+
+def format_af(af):
+    """Format alternate function.
+
+    Args:
+        af: Alternate function ("analog" or AF number).
+
+    Returns:
+        DT AF definition.
+    """
+
+    if af == "analog":
+        return "ANALOG"
+
+    return f"AF{af:d}"
+
+
+def format_drive(drive):
+    """Format drive mode.
+
+    Args:
+        drive: Drive mode ("push-pull", "open-drain").
+
+    Returns:
+        DT drive definition.
+    """
+
+    if drive not in ("push-pull", "open-drain"):
+        raise ValueError(f"Unsupported drive mode: {drive}")
+
+    return "drive-" + drive
+
+
+def format_bias(bias):
+    """Format bias mode.
+
+    Args:
+        bias: Bias mode ("disable", "pull-up", "pull-down").
+
+    Returns:
+        DT bias definition.
+    """
+
+    if bias not in ("disable", "pull-up", "pull-down"):
+        raise ValueError(f"Unsupported bias mode: {bias}")
+
+    return "bias-" + bias
+
+
+def format_remap(remap):
+    """Format remap number for DT.
+
+    Args:
+        remap: Remap number (0..3) or "analog".
+
+    Returns:
+        DT remap definition.
+    """
+
+    if remap == "analog" or remap == 0:
+        return "NO_REMAP"
+    elif remap == 1:
+        return "REMAP_1"
+    elif remap == 2:
+        return "REMAP_2"
+    elif remap == 3:
+        return "FULL_REMAP"
+
+    raise ValueError(f"Unsupported remap: {remap}")
+
+
+def format_mode(mode):
+    """Format mode for DT (F1 series).
+
+    Applied rules:
+        - Analog: Analog mode (ANALOG)
+        - Alternate function input: Input mode (GPIO_IN)
+        - Alternate function output: Alternate Function Output (ALTERNATE)
+        - Alternate function input/output: Alternate Function Output (ALTERNATE)
```

Comment:
For ease of use, it would be good to be closer to F1 bindings here:
```
            * 0 : Alternate function output
            * 1 : General purpose output
            * 2 : Input
            * 3 : Analog
```

Reply:
This function maps F1 mode depending on operation mode, which follows the rules mentioned in the comment of the function.